perception/traffic_light_arbiter/src/signal_match_validator.cpp

```diff
@@ -0,0 +1,274 @@
+    return stamp1;
+  } else {
+    return stamp2;
+  }
+}
+
+}  // namespace util
+
+autoware_perception_msgs::msg::TrafficSignalArray SignalMatchValidator::validateSignals(
+  const TrafficSignalArray & perception_signals, const TrafficSignalArray & external_signals)
+{
+  TrafficSignalArray validated_signals;
+
+  // Set newer stamp
+  validated_signals.stamp = util::get_newer_stamp(perception_signals.stamp, external_signals.stamp);
+
+  // Create the unique set of the received id,
+  // then compare the signal element for each received signal id
+  const auto received_signal_id_set =
+    util::create_signal_id_set(perception_signals.signals, external_signals.signals);
+
+  for (const auto & signal_id : received_signal_id_set) {
+    const auto perception_result = util::find_signal_by_id(perception_signals, signal_id);
+    const auto external_result = util::find_signal_by_id(external_signals, signal_id);
+
+    // Both results doesn't exist
+    if (!perception_result && !external_result) {
+      continue;
+    }
+
+    // We don't validate the pedestrian signals
+    // TODO(TomohitoAndo): Validate pedestrian signals
+    if (isPedestrianSignal(signal_id)) {
+      validated_signals.signals.emplace_back(util::get_highest_confidence_signal(
+        perception_result, external_result, external_priority_));
+
+      continue;
+    }
+
+    // If either of the signal is not received, treat as unknown signal
+    if (!perception_result && external_result) {
+      const auto unknown_signal = util::create_unknown_signal(*external_result);
+      validated_signals.signals.emplace_back(unknown_signal);
+      continue;
+    }
+    if (!external_result && perception_result) {
+      const auto unknown_signal = util::create_unknown_signal(*perception_result);
+      validated_signals.signals.emplace_back(unknown_signal);
+      continue;
+    }
+
+    // Check if they have the same elements
+    if (!util::are_all_elements_equivalent(
+          perception_result->elements, external_result->elements)) {
+      const auto unknown_signal = util::create_unknown_signal(*perception_result, *external_result);
+      validated_signals.signals.emplace_back(unknown_signal);
+      continue;
+    }
+
+    // Both results are same, then insert the received color
+    validated_signals.signals.emplace_back(*perception_result);
+  }
+
+  return validated_signals;
+}
+
+void SignalMatchValidator::setPedestrianSignals(
+  const std::vector<TrafficLightConstPtr> & pedestrian_signals)
+{
+  for (const auto & signal : pedestrian_signals) {
+    map_pedestrian_signal_regulatory_elements_set_.emplace(signal->id());
+  }
+}
+
+void SignalMatchValidator::setExternalPriority(const bool external_priority)
+{
+  external_priority_ = external_priority;
+}
+
+bool SignalMatchValidator::isPedestrianSignal(const lanelet::Id & signal_id)
+{
+  return map_pedestrian_signal_regulatory_elements_set_.find(signal_id) !=
```

Comment:
nit: this is O(n). I am not sure if given the dimensionality it matters or not but better to point it out

Reply:
Thank you for your suggestion! I googled about the time complexity of unordered_set.
`map_pedestrian_signal_regulatory_elements_set_` is `std::unordered_set`, which has a time complexity of O(1) on average, and O(n) on worst case. On the other hands, `std::set` has a complexity of O(log(n)). 
So your suggestion is to replace `unordered_set` with `set` to avoid the worst case?